### DISCUSSION TITLE
Admin search queries are extremely resource intensive

### DISCUSSION THREAD

**Discussion by paweltatarczuk** [craftcms/cms#8915] [Ideas]
### Description

We are experiencing major stability issues with the database after using search functionality in the admin panel. Search queries cause memory peaks and sometimes database crashes.

I've done some research and it all comes down to how Craft is performing a fulltext search, there are two queries being executed one by one:

1. Select all element IDs matching the search filters (not deleted, siteId, not archived, not draft, not revision)
2. Match elements by keyword (keyword, ids in array of elements id form previous query)

Sample queries:

```
SELECT `elements`.`id` FROM (
    SELECT `elements`.`id` AS `elementsId`, `elements_sites`.`id` AS `elementsSitesId`, `content`.`id` AS `contentId`, `structureelements`.`structureId`
    FROM `elements` `elements`
    INNER JOIN `entries` `entries` ON `entries`.`id` = `elements`.`id`
    INNER JOIN `elements_sites` `elements_sites` ON `elements_sites`.`elementId` = `elements`.`id`
    INNER JOIN `content` `content` ON (`content`.`elementId` = `elements`.`id`) AND (`content`.`siteId` = `elements_sites`.`siteId`)
    LEFT JOIN `structureelements` `structureelements` ON (`structureelements`.`elementId` = `elements`.`id`) AND (EXISTS (
        SELECT *
        FROM `structures`
        WHERE (`id` = `structureelements`.`structureId`) AND (`dateDeleted` IS NULL)
    ))
    WHERE (`entries`.`sectionId` IN ('5', '11', '8', '1', '7', '10', '2', '9', '4')) AND (`entries`.`sectionId` IN ('5', '11', '8', '1', '7', '10', '2', '9', '4')) AND (`elements_sites`.`siteId`=2) AND (`elements`.`archived`=FALSE) AND (`elements`.`dateDeleted` IS NULL) AND (`elements`.`draftId` IS NULL) AND (`elements`.`revisionId` IS NULL)
) `subquery`
INNER JOIN `entries` `entries` ON `entries`.`id` = `subquery`.`elementsId`
INNER JOIN `elements` `elements` ON `elements`.`id` = `subquery`.`elementsId`
INNER JOIN `elements_sites` `elements_sites` ON `elements_sites`.`id` = `subquery`.`elementsSitesId`
INNER JOIN `content` `content` ON `content`.`id` = `subquery`.`contentId`
LEFT JOIN `structureelements` `structureelements` ON (`structureelements`.`elementId` = `subquery`.`elementsId`) AND (`structureelements`.`structureId` = `subquery`.`structureId`)

SELECT * FROM `searchindex` WHERE MATCH(`keywords`) AGAINST('+needle*' IN BOOLEAN MODE) AND `siteId` = 2 AND `elementId` IN (1, 2, 3, ...)
```

I see two issues here:

- First query is fetching thousands of elements with a subquery
- Second query is passing thousands of element IDs in a where condition

Considering the same search functionality is being used in the Commerce plugin where number of orders increases over time the way search is executed kills the database performance. Database cannot be scaled up infinitely.

### Steps to reproduce

1. Enter "Entries" section in control panel
2. Search for a "needle"

### Additional info

- Craft version: 3.5.5 
- PHP version: 7.2.24-0ubuntu0.18.04.6
- Database driver & version: MySQL 8.0.20 (Managed MySQL by DigitalOcean)
- Plugins & versions:

**Comment by gojira32** [comment]
We've scaled our DB a couple of times already to combat this issue. I don't think it is feasible to keep doing so.

**Comment by paweltatarczuk** [comment]
It seems that MATCH ... AGAINST query is not using the FULLTEXT index (we have `subLeft` disabled):

```
SELECT * FROM `searchindex` WHERE MATCH(`keywords`) AGAINST('+foo*' IN BOOLEAN MODE) AND `siteId` = 2;
399 rows in set (18.74 sec)

SELECT * FROM `searchindex` WHERE keywords LIKE 'foo%' AND `siteId` = 2;
Empty set (3.89 sec)
```

MATCH ... AGAINST is only used for MySQL queries, shouldn't it use "LIKE" condition when `subLeft` is disabled?

---

EDIT: I can switch to "LIKE" condition with this config:

```
'defaultSearchTermOptions' => [
  'subLeft' => false, 
  'subRight' => true, 
  'exact' => true,
],
```

**For anyone wondering why is the config above important:**

FULLTEXT indexes can only be used for "starts with" queries:

https://stackoverflow.com/questions/2481528/mysql-like-performance-boost

**Comment by paweltatarczuk** [comment]
Unfortunately after making sure Craft is performing a search query that is actually using FULLTEXT index on `searchindex` table the initial subquery and massive `IN (...)` condition still remains an issue.

**Comment by brandonkelly** [comment]
We are aware of the scalability issues with this, and working to make it much better in Craft 4 :)
